### ISSUE TITLE
Gradient descent in the body of while loop (v2) causes error

### ISSUE THREAD

**Issue by samdow** [tensorflow/tensorflow#24866]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS 10.13.6, TensorFlow r1.12
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): v1.12.0-rc0
- Python version: 2.7.15
- Bazel version (if compiling from source): 0.20.0
- GCC/Compiler version (if compiling from source): 10.0.0
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
When using `tf.while_loop` (v1), we are able to call gradients in the body of the while loop. In other words, we were able to embed the training into our TensorFlow graph.

If we replace `tf.while_loop` with `while_v2.while_loop` then we end up getting an error:

`In op 'train/update_W/ApplyGradientDescent', input types ([tf.float32, tf.float32, tf.float32]) are not compatible with expected types ([tf.float32_ref, tf.float32, tf.float32])`

Since this shows that gradient descent is being fed a list of length 3, but the placeholders have a length of 2, I think it might be something around the counter getting added. However, I'm not certain and any insights you guys have would be useful!

**Describe the expected behavior**
We expect to get the same results when we use v1 or v2. We know that you guys are working on v2 of the while loops, and we're really excited for it. We just wanted to make certain this was on your radar as a use case, since we didn't see it in the RFC 😄 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
def body(loopVar):
    y = x * W
    loss = tf.reduce_mean((y - y_)**2)
    optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.1)
    train_op = optimizer.minimize(loss, name='train')

    with tf.control_dependencies([train_op]):
        return loopVar + 1

x = tf.placeholder(tf.float32, name='input')
y_ = tf.placeholder(tf.float32, name='target')
W = tf.Variable(5., name='W')

counter = tf.Variable(0, name='counter')
z = while_v2.while_loop(lambda v: v < 6, body, [counter])

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    print(sess.run(W))
    print(sess.run(z, feed_dict={x: [1.0, 2.0], y_:[2.0, 4.0]}))
    print(sess.run(W))
```

(If we replace `while_v2.while_loop` with `tf.while_loop` then we get the expected results of seeing the value of W move towards 2)

**Other info / logs**

**Comment by skye** [comment]
while_v2 currently works only with ResourcesVariables, which aren't enabled by default in TF yet. Try calling `tf.enable_resource_variables()` at the beginning of your code or passing `use_resource=True` to the `tf.Variable` constructor. Please see https://stackoverflow.com/questions/40817665/whats-the-difference-between-variable-and-resourcevariable-in-tensorflow for more information on ResourceVariables.

**Comment by samdow** [comment]
It worked! Thanks for your help!
